sources/sqlite3_cql_extension/cqlsqlite3extension.py

```diff
@@ -0,0 +1,355 @@
+
+
+    print("  // 6. Cleanup In arguments since they are no longer needed")
+    for arg in [arg for arg in in_arguments if is_ref_type[arg['type']]]:
+        if   arg['type'] == "text":   print(f"  cql_string_release({arg['name']});")
+        elif arg['type'] == "blob":   print(f"  cql_blob_release({  arg['name']});")
+        elif arg['type'] == "object": print(f"  cql_object_release({arg['name']});")
+    print("")
+
+
+    print("  // 7. Ensure the procedure executed successfully")
+    if proc['usesDatabase']:
+        print("  if (rc != SQLITE_OK) {")
+        print("    sqlite3_result_null(context);")
+        print("    goto cleanup;")
+        print("  }")
+    print("")
+
+
+    print("  // 8. Resolve the result base on:")
+    print("  //   (A) The first column of first row of the result_set, if any")
+    print("  //   (B) The first outtie argument (in or inout) value, if any")
+    print("  //   (C) Fallback to: null")
+    print("  //")
+    if proc['projection']:
+        print("  // Current strategy: (A) Using the result set")
+
+        print("  set_sqlite3_result_from_result_set(context, (cql_result_set_ref)_data_result_set_);")
+        print("  goto cleanup;")
+        print("")
+    else:
+        print("  // Current strategy: (B) Using Outtie arguments")
+        print("  // Set Sqlite result")
+        print("  // NB: If the procedure generates a cql result set, the first column of the first row would be used as the result")
+        for arg in [arg for arg in proc['args'] if arg.get('binding', 'in') in ("inout", "out")]:
+            skip = False
+
+            if arg['type'] == "text":
+                print(f"  SQLITE3_RESULT_CQL_TEXT(context, {arg['name']}); goto cleanup;")
+            elif arg['type'] == "blob":
+                print(f"  SQLITE3_RESULT_CQL_BLOB(context, {arg['name']}); goto cleanup;")
+            elif is_ref_type[arg['type']]:
+                # Object
+                print(f"  /* {arg['type']} not implemented yet */")
+                skip = True
+            elif arg['isNotNull']:
+                print(f"  {sqlite3_result_setter_notnull[arg['type']]}(context, {arg['name']}); goto cleanup;")
+            elif not arg['isNotNull']:
+                print(f"  {sqlite3_result_setter_nullable[arg['type']]}(context, {arg['name']}); goto cleanup;")
+            else:
+                print(f"  /* Unsupported type {arg['type']} */")
+                skip = True
+
+            if not skip:
+                break
+        print("")
+        print("  goto cleanup;")
+
+
+    print("invalid_arguments:")
+    print("  sqlite3_result_error(context, \"CQL extension: Invalid procedure arguments\", -1);")
+    print("  return;")
+
+
+    print("")
+    print("cleanup:")
+    print("  // 10. Cleanup Outtie arguments")
+    for arg in [arg for arg in outtie_arguments if is_ref_type[arg['type']]]:
+        if   arg['type'] == "text":   print(f"  if({arg['name']}) cql_string_release({arg['name']});")
```

Comment:
You don't need the null check here.  The release thingies all know to do a no-op if you are trying to release null.  This saves a lot of codegen.

Reply:
will try later